airlock/file_browser_api.py

```diff
@@ -133,7 +133,7 @@ def html_classes(self):
         distinguish file/dirs, and maybe even file types, in the UI, in case we
         need to.
         """
-        classes = [self.type.value.lower()]
+        classes = [self.type.value.lower()] if self.type else []
```

Comment:
AFAICT `self.type` is only null during construction of a `PathItem` in `build_path_tree`.  Adding this check to propitiate mypy is a bit of a smell -- instead can we make `type` required, and tweak `build_path_tree`?

Reply:
It might be possible. There's some awkwardness in PathItem construction, in that you have to create a parent PathItem before you create its children, so that you can pass a reference to the parent PathItem. But it may be that we can always know the type when we create the object, I'll have a look

Reply:
Yeah, it can work. Means branching twice, but ends up with just children being set after instance creation.
```
diff --git airlock/file_browser_api.py airlock/file_browser_api.py
index 8aa61eb..633fadf 100644
--- airlock/file_browser_api.py
+++ airlock/file_browser_api.py
@@ -34,8 +34,8 @@ class PathItem:
 
     container: AirlockContainer
     relpath: UrlPath
+    type: PathType
 
-    type: PathType = None
     children: list["PathItem"] = field(default_factory=list)
     parent: "PathItem" = None
 
@@ -346,13 +346,6 @@ def get_path_tree(
         for child, descendants in grouped.items():
             path = parent.relpath / child
             selected = path == selected_path
-            node = PathItem(
-                container=container,
-                relpath=path,
-                parent=parent,
-                selected=selected,
-                supporting_file=container.is_supporting_file(path),
-            )
 
             # If it has decendants, it is a directory. However, an empty
             # directory in workspace still needs to be classed as a PathType.DIR.
@@ -365,10 +358,22 @@ def get_path_tree(
             # is_dir(), but we are trying to avoid hitting the filesystem in
             # the tree recursion for speed.
             #
-            # We have a backstop check to not blow up in this case
+            # We have a backstop check to not blow up in this case in
             # PathItem.contents()
-            if descendants or path.suffix == "":
-                node.type = PathType.DIR
+            node_type = (
+                PathType.DIR if (descendants or path.suffix == "") else PathType.FILE
+            )
+
+            node = PathItem(
+                container=container,
+                relpath=path,
+                type=node_type,
+                parent=parent,
+                selected=selected,
+                supporting_file=container.is_supporting_file(path),
+            )
+
+            if node_type == PathType.DIR:
                 # recurse down the tree
                 node.children = build_path_tree(descendants, parent=node)
 
@@ -377,8 +382,6 @@ def get_path_tree(
                     node.expanded = True
                 else:
                     node.expanded = selected or (path in (selected_path.parents or []))
-            else:
-                node.type = PathType.FILE
 
             tree.append(node)
 
```